simulator/src/main/java/peersim/kademlia/das/operations/SamplingOperation.java

```diff
@@ -92,70 +106,111 @@ public BigInteger getRadiusNonValidator() {
-
-    List<BigInteger> nextNodes = new ArrayList<>();
-
-    BigInteger nextNode;
-    do {
-      nextNode = getNeighbour();
-      if (nextNode != null) nextNodes.add(nextNode);
-    } while (nextNode != null);
-    if (nextNodes.size() > 0) return nextNodes.toArray(new BigInteger[0]);
-    return new BigInteger[0];
-  }
-
-  public BigInteger getNeighbour() {
-
-    BigInteger res = null;
-    if (closestSet.isEmpty()) {
-      List<BigInteger> nodes = new ArrayList<>();
-      for (BigInteger sample : samples.keySet()) {
-
-        if (!samples.get(sample)) {
-          List<BigInteger> validatorsBySample =
-              searchTable.getValidatorNodesbySample(sample, radiusValidator);
-          List<BigInteger> nonValidatorsBySample =
-              searchTable.getNonValidatorNodesbySample(sample, radiusNonValidator);
-
-          boolean found = false;
-          if (validatorsBySample != null && validatorsBySample.size() > 0) {
-            nodes.addAll(validatorsBySample);
-            Collections.shuffle(nodes);
-            found = true;
+  protected void createNodes() {
+    for (BigInteger sample : samples.keySet()) {
+      if (!samples.get(sample).isDownloaded()) {
+
+        List<BigInteger> validatorsBySample = SearchTable.getNodesBySample(sample);
+        List<BigInteger> nonValidatorsBySample =
+            searchTable.getNonValidatorNodesbySample(sample, radiusNonValidator);
+
+        boolean found = false;
+        if (validatorsBySample != null && validatorsBySample.size() > 0) {
+          for (BigInteger id : validatorsBySample) {
+            if (!nodes.containsKey(id)) {
+              nodes.put(id, new Node(id));
+              nodes.get(id).addSample(samples.get(sample));
+            } else {
+              nodes.get(id).addSample(samples.get(sample));
+            }
           }
-          if (nonValidatorsBySample != null && nonValidatorsBySample.size() > 0) {
-            nodes.addAll(nonValidatorsBySample);
-            found = true;
+          found = true;
+        }
+        if (nonValidatorsBySample != null && nonValidatorsBySample.size() > 0) {
+          for (BigInteger id : nonValidatorsBySample) {
+            if (!nodes.containsKey(id)) {
+              nodes.put(id, new Node(id));
+              nodes.get(id).addSample(samples.get(sample));
+            } else {
+              nodes.get(id).addSample(samples.get(sample));
+            }
           }
-          if (!found && callback != null) callback.missing(sample, this);
+          found = true;
         }
+
+        if (!found && callback != null) callback.missing(sample, this);
       }
-      // Collections.shuffle(nodes);
+    }
+  }
+
+  public BigInteger[] doSampling() {
 
-      for (BigInteger node : nodes) {
-        if (closestSet.get(node) == null) {
-          closestSet.put(node, true);
+    aggressiveness += KademliaCommonConfigDas.aggressiveness_step;
+    for (Node n : nodes.values()) n.setAgressiveness(aggressiveness);
```

Comment:
Are node all the node the sampling peer knows?

Reply:
the node structure represents the known nodes for wanted samples to track asking nodes, num of samples provided by node and aggressiveness, for the fetching algoritthm